### ISSUE TITLE
MC inputsOutputs and validate endpoints take ~ 5 sec

### ISSUE THREAD

**Issue by jmthibault79** [DataBiosphere/firecloud-app#135]
During the Testing Party on Staging we noticed that viewing a Method Config is very slow.  Looking at the Network tab, we saw that these 2 endpoints take about 5 sec each.  Can we reduce this?

**Comment by rtitle** [comment]
Here is a strack trace taken from Rawls while the MC was being loaded:
 
```
"scala-execution-context-global-59" #59 daemon prio=5 os_prio=0 tid=0x00007f842c049000 nid=0x49 runnable [0x00007f847c53f000]
   java.lang.Thread.State: RUNNABLE
	at scala.runtime.Statics.anyHash(Statics.java:115)
	at scala.util.hashing.MurmurHash3.productHash(MurmurHash3.scala:64)
	at scala.util.hashing.MurmurHash3$.productHash(MurmurHash3.scala:211)
	at scala.runtime.ScalaRunTime$._hashCode(ScalaRunTime.scala:145)
	at scala.Tuple2.hashCode(Tuple2.scala:19)
	at scala.runtime.Statics.anyHash(Statics.java:115)
	at scala.util.hashing.MurmurHash3.$anonfun$unorderedHash$1(MurmurHash3.scala:92)
	at scala.util.hashing.MurmurHash3.$anonfun$unorderedHash$1$adapted(MurmurHash3.scala:91)
	at scala.util.hashing.MurmurHash3$$Lambda$2613/514206257.apply(Unknown Source)
	at scala.collection.immutable.HashMap$HashMap1.foreach(HashMap.scala:231)
	at scala.collection.immutable.HashMap$HashTrieMap.foreach(HashMap.scala:462)
	at scala.collection.immutable.HashMap$HashTrieMap.foreach(HashMap.scala:462)
	at scala.collection.immutable.HashMap$HashTrieMap.foreach(HashMap.scala:462)
	at scala.util.hashing.MurmurHash3.unorderedHash(MurmurHash3.scala:91)
	at scala.util.hashing.MurmurHash3$.mapHash(MurmurHash3.scala:225)
	at scala.collection.GenMapLike.hashCode(GenMapLike.scala:35)
	at scala.collection.GenMapLike.hashCode$(GenMapLike.scala:35)
	at scala.collection.AbstractMap.hashCode(Map.scala:59)
	at scala.runtime.Statics.anyHash(Statics.java:115)
	at scala.util.hashing.MurmurHash3.productHash(MurmurHash3.scala:64)
	at scala.util.hashing.MurmurHash3$.productHash(MurmurHash3.scala:211)
	at scala.runtime.ScalaRunTime$._hashCode(ScalaRunTime.scala:145)
	at wdl.draft2.model.WdlSyntaxErrorFormatter.hashCode(WdlSyntaxErrorFormatter.scala:10)
	at scala.runtime.Statics.anyHash(Statics.java:115)
	at scala.util.hashing.MurmurHash3.productHash(MurmurHash3.scala:64)
	at scala.util.hashing.MurmurHash3$.productHash(MurmurHash3.scala:211)
	at scala.runtime.ScalaRunTime$._hashCode(ScalaRunTime.scala:145)
	at wdl.draft2.model.WdlWorkflow.hashCode(WdlWorkflow.scala:46)
	at scala.runtime.Statics.anyHash(Statics.java:115)
	at scala.util.hashing.MurmurHash3.productHash(MurmurHash3.scala:64)
	at scala.util.hashing.MurmurHash3$.productHash(MurmurHash3.scala:211)
	at scala.runtime.ScalaRunTime$._hashCode(ScalaRunTime.scala:145)
	at scala.Some.hashCode(Option.scala:333)
	at scala.runtime.Statics.anyHash(Statics.java:115)
	at scala.util.hashing.MurmurHash3.productHash(MurmurHash3.scala:64)
	at scala.util.hashing.MurmurHash3$.productHash(MurmurHash3.scala:211)
	at scala.runtime.ScalaRunTime$._hashCode(ScalaRunTime.scala:145)
	at wdl.draft2.model.Declaration.hashCode(Declaration.scala:164)
	at scala.runtime.Statics.anyHash(Statics.java:115)
	at scala.collection.immutable.HashSet.elemHashCode(HashSet.scala:178)
	at scala.collection.immutable.HashSet.computeHash(HashSet.scala:187)
	at scala.collection.immutable.HashSet.$plus(HashSet.scala:83)
	at scala.collection.immutable.HashSet.$plus(HashSet.scala:34)
	at scala.collection.SetLike.$anonfun$$plus$plus$1(SetLike.scala:150)
	at scala.collection.SetLike$$Lambda$57/757332719.apply(Unknown Source)
	at scala.collection.TraversableOnce.$anonfun$foldLeft$1(TraversableOnce.scala:157)
	at scala.collection.TraversableOnce.$anonfun$foldLeft$1$adapted(TraversableOnce.scala:157)
	at scala.collection.TraversableOnce$$Lambda$51/1429462130.apply(Unknown Source)
	at scala.collection.immutable.HashSet$HashSet1.foreach(HashSet.scala:320)
	at scala.collection.immutable.HashSet$HashTrieSet.foreach(HashSet.scala:976)
	at scala.collection.TraversableOnce.foldLeft(TraversableOnce.scala:157)
	at scala.collection.TraversableOnce.foldLeft$(TraversableOnce.scala:155)
	at scala.collection.AbstractTraversable.foldLeft(Traversable.scala:104)
	at scala.collection.TraversableOnce.$div$colon(TraversableOnce.scala:151)
	at scala.collection.TraversableOnce.$div$colon$(TraversableOnce.scala:151)
	at scala.collection.AbstractTraversable.$div$colon(Traversable.scala:104)
	at scala.collection.SetLike.$plus$plus(SetLike.scala:150)
	at scala.collection.SetLike.$plus$plus$(SetLike.scala:150)
	at scala.collection.AbstractSet.$plus$plus(Set.scala:47)
	at wdl.draft2.model.WdlGraphNode$.wdl$draft2$model$WdlGraphNode$$calculateUpstreamAncestry(WdlGraphNode.scala:52)
	at wdl.draft2.model.WdlGraphNode$.$anonfun$calculateUpstreamAncestry$1(WdlGraphNode.scala:54)
	at wdl.draft2.model.WdlGraphNode$$$Lambda$2912/188099430.apply(Unknown Source)
	at scala.collection.TraversableOnce.$anonfun$foldLeft$1(TraversableOnce.scala:157)
	at scala.collection.TraversableOnce.$anonfun$foldLeft$1$adapted(TraversableOnce.scala:157)
	at scala.collection.TraversableOnce$$Lambda$51/1429462130.apply(Unknown Source)
	at scala.collection.immutable.Set$Set2.foreach(Set.scala:130)
	at scala.collection.TraversableOnce.foldLeft(TraversableOnce.scala:157)
	at scala.collection.TraversableOnce.foldLeft$(TraversableOnce.scala:155)
	at scala.collection.AbstractTraversable.foldLeft(Traversable.scala:104)
	at wdl.draft2.model.WdlGraphNode$.wdl$draft2$model$WdlGraphNode$$calculateUpstreamAncestry(WdlGraphNode.scala:54)
	at wdl.draft2.model.WdlGraphNode.upstreamAncestry(WdlGraphNode.scala:12)
	at wdl.draft2.model.WdlGraphNode.upstreamAncestry$(WdlGraphNode.scala:12)
	at wdl.draft2.model.Declaration.upstreamAncestry$lzycompute(Declaration.scala:164)
	- locked <0x00000000f5fb8aa0> (a wdl.draft2.model.Declaration)
	at wdl.draft2.model.Declaration.upstreamAncestry(Declaration.scala:164)
	at wdl.draft2.model.DeclarationInterface.asWorkflowInput(Declaration.scala:80)
	at wdl.draft2.model.DeclarationInterface.asWorkflowInput$(Declaration.scala:74)
	at wdl.draft2.model.Declaration.asWorkflowInput(Declaration.scala:164)
	at wdl.draft2.model.WdlCall.$anonfun$workflowInputs$2(WdlCall.scala:87)
	at wdl.draft2.model.WdlCall$$Lambda$2920/1737919424.apply(Unknown Source)
	at scala.collection.TraversableLike.$anonfun$flatMap$1(TraversableLike.scala:241)
	at scala.collection.TraversableLike$$Lambda$47/947647193.apply(Unknown Source)
	at scala.collection.Iterator.foreach(Iterator.scala:929)
	at scala.collection.Iterator.foreach$(Iterator.scala:929)
	at scala.collection.AbstractIterator.foreach(Iterator.scala:1417)
	at scala.collection.IterableLike.foreach(IterableLike.scala:71)
	at scala.collection.IterableLike.foreach$(IterableLike.scala:70)
	at scala.collection.AbstractIterable.foreach(Iterable.scala:54)
	at scala.collection.TraversableLike.flatMap(TraversableLike.scala:241)
	at scala.collection.TraversableLike.flatMap$(TraversableLike.scala:238)
	at scala.collection.AbstractTraversable.flatMap(Traversable.scala:104)
	at wdl.draft2.model.WdlCall.workflowInputs(WdlCall.scala:87)
	at wdl.draft2.model.WdlWorkflow.$anonfun$unsatisfiedCallInputs$1(WdlWorkflow.scala:61)
	at wdl.draft2.model.WdlWorkflow$$Lambda$2918/2142124653.apply(Unknown Source)
	at scala.collection.TraversableLike.$anonfun$flatMap$1(TraversableLike.scala:241)
	at scala.collection.TraversableLike$$Lambda$47/947647193.apply(Unknown Source)
	at scala.collection.immutable.HashSet$HashSet1.foreach(HashSet.scala:320)
	at scala.collection.immutable.HashSet$HashTrieSet.foreach(HashSet.scala:976)
	at scala.collection.TraversableLike.flatMap(TraversableLike.scala:241)
	at scala.collection.TraversableLike.flatMap$(TraversableLike.scala:238)
	at scala.collection.AbstractTraversable.flatMap(Traversable.scala:104)
	at wdl.draft2.model.WdlWorkflow.unsatisfiedCallInputs(WdlWorkflow.scala:61)
	at wdl.draft2.model.WdlWorkflow.inputs(WdlWorkflow.scala:76)
	at org.broadinstitute.dsde.rawls.jobexec.MethodConfigResolver$.$anonfun$getMethodInputsOutputs$1(MethodConfigResolver.scala:154)
	at org.broadinstitute.dsde.rawls.jobexec.MethodConfigResolver$$$Lambda$2989/2016746447.apply(Unknown Source)
	at scala.util.Success.$anonfun$map$1(Try.scala:251)
	at scala.util.Success.map(Try.scala:209)
	at org.broadinstitute.dsde.rawls.jobexec.MethodConfigResolver$.getMethodInputsOutputs(MethodConfigResolver.scala:152)
	at org.broadinstitute.dsde.rawls.workspace.WorkspaceService.$anonfun$getMethodInputsOutputs$3(WorkspaceService.scala:1398)
	at org.broadinstitute.dsde.rawls.workspace.WorkspaceService$$Lambda$2988/2092380610.apply(Unknown Source)
	at org.broadinstitute.dsde.rawls.util.MethodWiths.withWdl(WithSupport.scala:41)
	at org.broadinstitute.dsde.rawls.util.MethodWiths.withWdl$(WithSupport.scala:38)
	at org.broadinstitute.dsde.rawls.workspace.WorkspaceService.withWdl(WorkspaceService.scala:50)
	at org.broadinstitute.dsde.rawls.workspace.WorkspaceService.$anonfun$getMethodInputsOutputs$2(WorkspaceService.scala:1398)
	at org.broadinstitute.dsde.rawls.workspace.WorkspaceService$$Lambda$2987/1944897763.apply(Unknown Source)
	at org.broadinstitute.dsde.rawls.util.MethodWiths.$anonfun$withMethod$1(WithSupport.scala:33)
	at org.broadinstitute.dsde.rawls.util.MethodWiths$$Lambda$2628/2042856483.apply(Unknown Source)
	at slick.basic.BasicBackend$DatabaseDef.$anonfun$runInContext$1(BasicBackend.scala:147)
	at slick.basic.BasicBackend$DatabaseDef$$Lambda$1783/2029512000.apply(Unknown Source)
	at scala.concurrent.Future.$anonfun$flatMap$1(Future.scala:302)
	at scala.concurrent.Future$$Lambda$465/125173444.apply(Unknown Source)
	at scala.concurrent.impl.Promise.$anonfun$transformWith$1(Promise.scala:37)
	at scala.concurrent.impl.Promise$$Lambda$466/209524543.apply(Unknown Source)
	at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:60)
	at scala.concurrent.impl.ExecutionContextImpl$AdaptedForkJoinTask.exec(ExecutionContextImpl.scala:140)
	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
	at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)
	at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
	at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157)
```

**Comment by jmthibault79** [comment]
It also seems to wait for validate to complete before running inputsOutputs.  Can we do them in parallel?

**Comment by jmthibault79** [comment]
We see a major slowdown after the 7/17 release.  That suggests https://github.com/broadinstitute/rawls/pull/945 has something do with it.

**Comment by jmthibault79** [comment]
https://github.com/broadinstitute/rawls/pull/968 improves performance after the first access of a WDL, but the first access will still be slow.  We should still investigate why.

**Comment by rebrown1395** [comment]
Won't fix.
